### PR TITLE
Add a `structseq` class to `_typeshed`

### LINKED ISSUE #6546: Weird `n_fields` attributes, etc.
There are a bunch of `NamedTuple`-like classes in some of the stranger corners of the stdlib that have the following three attributes:

- `n_fields`
- `n_sequence_fields`
- `n_unnamed_fields`

For example:

https://github.com/python/typeshed/blob/1fdd7e40afbad567a92de4ce98bdddaa6be14db5/tests/stubtest_allowlists/py3_common.txt#L620-L631

@Akuli [counts 22 classes](https://github.com/python/typeshed/pull/6500#discussion_r762599896) that are missing these attributes. Some classes, however, _do_ currently define these attributes in the stub currently. `time._struct_time` defines them as properties (which seems incorrect -- they seem more like `ClassVar`s to me):

https://github.com/python/typeshed/blob/1fdd7e40afbad567a92de4ce98bdddaa6be14db5/stdlib/time.pyi#L35-L50

As discussed [here](https://github.com/python/typeshed/pull/6500#discussion_r762599896), these attributes don't seem particularly _useful_. I'm also not sure there's much _harm_ in having them in the stubs, but it might be weird to have them popping up in autocompletion suggestions.

In any event, they should probably either all be added to the stubs, or all removed from stubs, and added to the "wontfix" section of the allowlists instead of the "missing from stubs" section of the allowlists.

Thoughts?

### PR DESCRIPTION
**Pull request by AlexWaygood**
Fixes #6546.

### CODE REVIEW

File: stdlib/_thread.pyi
Comment (Akuli) [inline stdlib/_thread.pyi:48]: This should probably return `Tuple[Any, ...]`. The same goes for other `__getitem__` definitions too.

File: stdlib/_typeshed/__init__.pyi
Comment (Akuli) [inline stdlib/_typeshed/__init__.pyi:210]: This inherits from tuple at runtime. It might be good to reflect that in stubs. Does this work?
```suggestion
class structseq(Tuple[_T_co, ...], Generic[_T_co]):
```

File: stdlib/pwd.pyi
Comment (Akuli) [inline stdlib/pwd.pyi:6]: I'm arfaid these comments might get outdated easily.

File: stdlib/time.pyi
Comment (Akuli) [inline stdlib/time.pyi:47]: We could use `int | Any`, which won't error if you assume it's an int, but also won't cause unreachable code warnings if you use `assert isinstance(thing, str)`.

File: stdlib/os/__init__.pyi
Comment (Akuli) [inline stdlib/os/__init__.pyi:920]: What does the `__new__` do? Have you checked that it works?

Comment (AlexWaygood) [reply stdlib/os/__init__.pyi:920]: I have checked that it works. This subclass of `structseq` appears to be unique, in that (presumably because it creates a tuple of length 1), the argument passed to the constructor does _not_ need to be an iterable. As such, I override the `structseq` constructor here.

File: stdlib/time.pyi
Comment (Akuli) [reply stdlib/time.pyi:47]: If we do this, you should no longer get warnings about overloads with non-overlapping return types, and we should be able to remove `# type: ignore[misc]`.

Comment (AlexWaygood) [reply stdlib/time.pyi:47]: Good shout.

File: stdlib/_typeshed/__init__.pyi
Comment (AlexWaygood) [reply stdlib/_typeshed/__init__.pyi]: Have made the change, thanks 👍

File: stdlib/pwd.pyi
Comment (Akuli) [reply stdlib/pwd.pyi:6]: But it doesn't error if the comment says 11, and somewhere 30 lines below there's a 12th member.

Comment (AlexWaygood) [reply stdlib/pwd.pyi:6]: Reckon I should delete them all, then, and just have one comment in the `structseq` definition? I worry people won't see it there if they're browsing the stubs, and think it's important to understand

**Comment by Akuli** [comment]
The mypy false positives from `zerver` are a bit unfortunate. Apparently mypy doesn't know that `foo[:4]` creates a tuple that is no more than 4 elements long, and assumes that all positional arguments get their values from that tuple. If I understood https://github.com/python/mypy/pull/11589 correctly, this is inconsistent with how mypy currently works with the `**mapping` syntax.

**Comment by JukkaL** [comment]
This will cause regressions, since the base classes change from a fixed-length tuple to `Tuple[x, ...]`. This can result in less precise type checking (if `x` is `Any`) and false positives. For example, these won't be compatible with a fixed-length tuple type, unlike previously. This also probably breaks some special casing in mypy that works for fixed-length tuples and named tuples. 

I have two ideas that should avoid the regressions:
1. Continue to define these as named tuples, but override things that are different (`__init__`); possibly inherit from a stub-only mixin class to add extra attributes such as `n_fields`.
2. Inherit from a fixed-length tuple, so that type checker special casing continues to work. Again possibly inherit from a mixin class to get `n_fields` without too much repetition.

**Comment by Akuli** [comment]
I like the fixed-length tuple idea. That way we don't need overloads like `def __getitem__(self, Literal[0])`.

**Comment by AlexWaygood** [comment]
> This will cause regressions, since the base classes change from a fixed-length tuple to `Tuple[x, ...]`. This can result in less precise type checking (if `x` is `Any`) and false positives. For example, these won't be compatible with a fixed-length tuple type, unlike previously. This also probably breaks some special casing in mypy that works for fixed-length tuples and named tuples.
> 
> I have two ideas that should avoid the regressions:
> 
> 1. Continue to define these as named tuples, but override things that are different (`__init__`); possibly inherit from a stub-only mixin class to add extra attributes such as `n_fields`.
> 2. Inherit from a fixed-length tuple, so that type checker special casing continues to work. Again possibly inherit from a mixin class to get `n_fields` without too much repetition.

Okay, I'll make revisions along these lines. Thanks!

**Comment by AlexWaygood** [comment]
> I like the fixed-length tuple idea. That way we don't need overloads like `def __getitem__(self, Literal[0])`.

Yes, especially with `_ExceptHookArgs`, we're not really gaining much by inheriting from `structseq` as it currently stands

**Comment by AlexWaygood** [comment]
> I have two ideas that should avoid the regressions

It has to be Option 2 (inherit from a fixed-length tuple), since there is no way to add a mutable `ClassVar` (or an immutable `ClassVar`) to a `NamedTuple`.

**Comment by AlexWaygood** [comment]
> Diff from [mypy_primer](https://github.com/hauntsaninja/mypy_primer), showing the effect of this PR on open source code:
> 
> ```diff
> mypy (https://github.com/python/mypy)
> + mypy/fscache.py:149: error: unused "type: ignore" comment
> ```

@JukkaL, looks like your suggestion got rid of the `zerver` false positives!

**Comment by AlexWaygood** [comment]
I have removed the following three attributes from the `os.stat_result` class, as they do not appear to be present on any system we test in CI, on any version of Python:

- `st_creator`
- `st_rsize`
- `st_type`

In the documentation, these three attributes are introduced with:

> On macOS systems, the following attributes may also be available:

However, that does not appear to be the case in reality.

File: stdlib/pwd.pyi
Comment (Akuli) [reply stdlib/pwd.pyi]: To me it seems like something that I would care about only if I wonder what a `structseq` is, so it belongs near `class structseq`. Most struct sequences don't have to override the constructor.

Comment (AlexWaygood) [reply stdlib/pwd.pyi]: Alright, I'll take them out, then.

**Comment by Akuli** [comment]
> On macOS systems, the following attributes may also be available:

Apparently this means an older macos that isn't really used anymore, if recent Pythons even work on it: https://stackoverflow.com/a/18779414

**Comment by AlexWaygood** [comment]
> > On macOS systems, the following attributes may also be available:
> 
> Apparently this means an older macos that isn't really used anymore, if recent Pythons even work on it: https://stackoverflow.com/a/18779414

Wow that's uhhh badly written documentation!

File: stdlib/_typeshed/__init__.pyi
Comment (Akuli) [inline stdlib/_typeshed/__init__.pyi:201]: The keys of the dict are strings:
```
>>> os.stat_result((1,2,3,4,5,6,7,8,9,10), {'st_rdev': 123}).st_rdev
123
```

```suggestion
    def __new__(cls: Type[_T], sequence: Iterable[_T_co], dict: dict[str, Any] = ...) -> _T: ...
```

File: stdlib/os/__init__.pyi
Comment (Akuli) [inline stdlib/os/__init__.pyi:337]: "Must be darwin" doesn't describe "other UNIX systems" very well.

```suggestion
    if sys.platform != "win32" and sys.platform != "linux":
```

File: stdlib/os/__init__.pyi
Comment (Akuli) [inline stdlib/os/__init__.pyi:404]: I don't think we need this comment.

Comment (AlexWaygood) [reply stdlib/os/__init__.pyi:404]: I didn't add the comment, was just preserving it from the existing stub. Happy to remove it, however.

Comment (AlexWaygood) [reply stdlib/os/__init__.pyi:404]: I agree it doesn't seem needed.

Comment (Akuli) [reply stdlib/os/__init__.pyi:404]: Let's remove it.

File: stdlib/time.pyi
Comment (AlexWaygood) [inline stdlib/time.pyi:40]: @JukkaL, I'm not sure inheriting from a fixed-length tuple is the right solution for this one specifically, since it can be a tuple of length 9, 10 or 11, depending on how long the argument passed to the constructor is. Maybe inheriting from a variable-length tuple would be better for this one specifically? Also interested in @Akuli's thoughts on this.

Comment (Akuli) [reply stdlib/time.pyi:40]: It seems to always behave like a tuple of 9 elements:

```
>>> list(time.struct_time((1,2,3,4,5,6,7,8,9,10,11)))
[1, 2, 3, 4, 5, 6, 7, 8, 9]
```

Comment (AlexWaygood) [reply stdlib/time.pyi:40]: Ahh, that simplifies things, thanks.

Comment (AlexWaygood) [reply stdlib/time.pyi]: Seems like the same goes for `os.stat_result`, too -- it always behaves like a tuple of length 10, even though the constructor takes an iterable of variable length.

File: stdlib/_typeshed/__init__.pyi
Comment (Akuli) [inline stdlib/_typeshed/__init__.pyi:204]: This part of the comment repeats what the type hint says. This also applies to using `dict` as opposed to `Mapping` or a protocol.

```suggestion
```

Avoiding unnecessary comments is hard, and it was hard for me too when I was new to typeshed. See the very tl;dr comments in `tkinter/__init__.pyi` for an example of how not to write comments in typeshed :)

**Review by Akuli — APPROVED** [review]
Thanks! I'll wait for another maintainer to merge this, in case someone else wants to review it too.

**Comment by AlexWaygood** [comment]
Friendly ping :)

Any chance of getting of getting this merged?

**Review by srittau — APPROVED** [review]

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Thanks, everybody! 😀

**Comment by AlexWaygood** [comment]
Thanks, everybody! 😀

**Pull request merged**
